### PR TITLE
Add new PostProcessingBlock to decouple post processing from LLMBlock

### PR DESCRIPTION
**Pull request by eshwarprasadS**
# Add PostProcessingBlock for Output Parsing

## Summary

This PR introduces a new `PostProcessingBlock` under `utilblocks.py` to handle output parsing, which was previously embedded in the `LLMBlock`. It also adds comprehensive tests for the new block to ensure robust parsing functionality and maintainability.

### New PostProcessingBlock

- **Implemented**: `PostProcessingBlock` in `src/sdg_hub/blocks/utilblocks.py`
  - Accepts `block_name`, `input_cols`, and `output_cols` as mandatory parameters (string or list).
  - Supports both tag-based and custom regex-based parsing.
  - Handles edge cases such as uneven tags, whitespace, and missing input columns.
  - Provides input validation with warnings for multiple input columns and errors for missing input columns.

### Tests

- **Added**: `tests/blocks/utilblocks/test_postprocessingblock.py`
  - Covers tag-based parsing, custom regex parsing, input validation, edge cases, and the `generate` method.
  - Ensures consistent and correct parsing behavior across a variety of scenarios.

### Low Risk

- **No Breaking Changes**: The new block is additive and does not modify existing block behavior.
- **Backward Compatibility**: Existing blocks and flows remain unaffected.

### Compatibility

- **Block API**: Follows the new design pattern for block initialization.
- **Test Coverage**: All new logic is covered by tests, ensuring reliability.

## Testing

- All new and existing tests pass.
- Edge cases and error handling are explicitly tested.

## Related Issues

- Part of the ongoing refactor to modularize output parsing logic (see discussion in #105).

## Notes

- The new block paves the way for further decoupling of parsing logic from generation blocks, improving code clarity and reusability.
- The test suite can serve as a reference for future block development.

## Summary by CodeRabbit

* **New Features**
  * Introduced a new string parsing tool for extracting and processing outputs from language models, supporting tag-based and regex-based extraction with optional cleanup.
  * Made the new string parser available as part of the public API.

* **Tests**
  * Added extensive tests to ensure correct parsing behavior, robust error handling, and coverage of various edge cases for the new string parsing tool.

### CODE REVIEW

File: src/sdg_hub/blocks/utilblocks.py [hunk 1]
```diff
@@ -595,3 +596,146 @@ def generate(self, samples: Dataset, **gen_kwargs: Dict[str, Any]) -> Dataset:
...
+        parser_name: Optional[str] = None,
+        parsing_pattern: Optional[str] = None,
+        parser_cleanup_tags: Optional[List[str]] = None,
+    ) -> None:
+        super().__init__(block_name)
+        self.input_cols = [input_cols] if isinstance(input_cols, str) else input_cols
+        self.output_cols = [output_cols] if isinstance(output_cols, str) else output_cols
+        self.start_tags = start_tags
+        self.end_tags = end_tags
+        self.parser_name = parser_name
```
Comment (shivchander) [inline src/sdg_hub/blocks/utilblocks.py]: we dont really have a need for `parser_name` - we can drop this

File: src/sdg_hub/blocks/utilblocks.py [hunk 2]
```diff
@@ -595,3 +596,146 @@ def generate(self, samples: Dataset, **gen_kwargs: Dict[str, Any]) -> Dataset:
...
+        self.parsing_pattern = parsing_pattern
+        self.parser_cleanup_tags = parser_cleanup_tags
+        self._validate()
+
+    def _validate(self) -> None:
+        """Validate the block configuration."""
+        # Validate custom parser configuration
+        if self.parser_name == "custom" and self.parsing_pattern is None:
```
Comment (shivchander) [inline src/sdg_hub/blocks/utilblocks.py]: this won't be needed if you address the `parser_name` issue above

File: src/sdg_hub/blocks/utilblocks.py [hunk 3]
```diff
@@ -595,3 +596,146 @@ def generate(self, samples: Dataset, **gen_kwargs: Dict[str, Any]) -> Dataset:
...
+        self.input_cols = [input_cols] if isinstance(input_cols, str) else input_cols
+        self.output_cols = [output_cols] if isinstance(output_cols, str) else output_cols
+        self.start_tags = start_tags
+        self.end_tags = end_tags
+        self.parser_name = parser_name
+        self.parsing_pattern = parsing_pattern
+        self.parser_cleanup_tags = parser_cleanup_tags
+        self._validate()
+
+    def _validate(self) -> None:
```
Comment (shivchander) [inline src/sdg_hub/blocks/utilblocks.py]: Could we drop the `_validate` for now? We might need to define a required `validate` method in the base Block abstract which does some standard data validation as well - Will create an issue for this

File: src/sdg_hub/blocks/utilblocks.py [hunk 4]
```diff
@@ -595,3 +596,146 @@ def generate(self, samples: Dataset, **gen_kwargs: Dict[str, Any]) -> Dataset:
...
+                self.start_tags,
+                self.end_tags,
+                self.output_cols,
+            ):
+                matches[output_col] = self._extract_matches(
+                    generated_string, start_tag, end_tag
+                )
+        return matches
+
+    def generate(self, samples: Dataset, **gen_kwargs: Dict[str, Any]) -> Dataset:
```
Comment (shivchander) [inline src/sdg_hub/blocks/utilblocks.py]: do we need gen_kwargs here? what do we use that for?

File: tests/blocks/utilblocks/test_postprocessingblock.py
Comment (shivchander) [inline tests/blocks/utilblocks/test_postprocessingblock.py:1]: Could we check if all the tests under `tests/blocks/test_llmblock.py` are captured here? If not could we adapt those as well? We can deprecate those tests soon

File: src/sdg_hub/blocks/utilblocks.py (hunk 4 above)
Comment (eshwarprasadS) [reply src/sdg_hub/blocks/utilblocks.py]: good catch, I repurposed the method signatures from existing blocks, must have been a remnant of that. Addressed!

**Review by shivchander — APPROVED** [review]
